optuna/study/study.py

```diff
@@ -154,7 +164,23 @@ def best_trial(self) -> FrozenTrial:
                 "using Study.best_trials to retrieve a list containing the best trials."
             )
 
-        return copy.deepcopy(self._storage.get_best_trial(self._study_id))
+        best_trial = self._storage.get_best_trial(self._study_id)
+
+        # If the trial with the best value is infeasible, select the best trial from all feasible
+        # trials. Note that the behavior is undefined when constrained optimization without the
+        # violation value in the best-valued trial.
+        constraints = best_trial.system_attrs.get(_CONSTRAINTS_KEY)
+        if constraints is not None and any([x > 0.0 for x in constraints]):
+            complete_trials = self.get_trials(deepcopy=False, states=[TrialState.COMPLETE])
```

Comment:
In my understanding, the cache in `_get_trials` is taken at the beginning of the trial, so it is not appropriate to use that cache in `best_trial`. It can be a strange value since `best_trial` is expected to be called after the trial.